src/storage-operators/src/s3_oneshot_sink.rs

```diff
@@ -96,126 +363,135 @@ pub fn copy_to<G, F>(
+        // NOTE: This is being refactored in https://github.com/MaterializeInc/materialize/pull/26489/
+        while let Some(event) = start_handle.next().await {
+            match event {
+                AsyncEvent::Data(cap, data) => {
+                    for res in data {
+                        if res.is_err() {
+                            completion.give(&cap, res.map(|_| 0)).await;
+                            return;
+                        }
                     }
                 }
+                AsyncEvent::Progress(_) => {}
             }
         }
 
-        // Map of an uploader per batch.
-        let mut s3_uploaders: BTreeMap<u64, CopyToS3Uploader> = BTreeMap::new();
-        let mut row_count = 0;
-        while let Some(event) = input_handle.next().await {
-            match event {
-                AsyncEvent::Data(_ts, data) => {
-                    for (((row, batch), ()), ts, diff) in data {
-                        if !up_to.less_equal(&ts) {
-                            if diff < 0 {
-                                onetime_callback(Err(format!(
-                                    "Invalid data in source errors, saw retractions ({}) for row that does not exist", diff * -1,
-                                )));
-                                return;
-                            }
-                            row_count += u64::try_from(diff).unwrap();
-                            let uploader = s3_uploaders
-                                .entry(batch)
-                                .or_insert_with(|| {
-                                    debug!("worker_id: {} will be handling batch: {}", worker_id, batch);
-                                    let file_name_prefix = format!("batch-{}-{:04}", &sink_id, batch);
-                                    CopyToS3Uploader::new(sdk_config.clone(), connection_details.clone(), file_name_prefix)
+        // fallible async block to use the `?` operator for convenience
+        let res = async move {
+            let sdk_config = aws_connection
+                .load_sdk_config(&connection_context, connection_id, InTask::Yes)
+                .await?;
+
+            // Map of an uploader per batch.
+            let mut s3_uploaders: BTreeMap<u64, CopyToS3Uploader> = BTreeMap::new();
+            let mut row_count = 0;
+            while let Some(event) = input_handle.next().await {
+                match event {
+                    AsyncEvent::Data(_ts, data) => {
+                        for (((row, batch), ()), ts, diff) in data {
+                            if !up_to.less_equal(&ts) {
+                                if diff < 0 {
+                                    Err(anyhow::anyhow!(
+                                        "Invalid data in source errors, saw retractions ({}) for \
+                                        row that does not exist",
+                                        diff * -1,
+                                    ))?;
```

Comment:
It's a bit strange to use the try operator on a thing that is explicitly an error. What you have here is what the bail macro does: `bail!("Invalid ...`

You might have seen instances of questions marks applied on err values in the pg source, in that case we do it because we are inside an `async_stream!` block which unfortunately does not understand `bail!` or other nested macros inside it..

Reply:
got it - switching to `anyhow::bail!`